compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts

```diff
@@ -346,6 +423,23 @@ function getReoderability(instr: Instruction): Reorderability {
     case "UnaryExpression": {
       return Reorderability.Reorderable;
     }
+    case "LoadLocal": {
+      const name = instr.value.place.identifier.name;
+      if (name !== null && name.kind === "named") {
+        const lastAssignment = references.lastAssignments.get(name.value);
+        const range = references.accessedRanges.get(instr.lvalue.identifier.id);
+        if (
+          lastAssignment !== undefined &&
+          lastAssignment < instr.id &&
+          range !== undefined &&
+          range.end === range.start // this LoadLocal is used exactly once
```

Comment:
I don't think I understand what the `accessedRanges` conditions are enforcing here, or why it matters that a local is used exactly once. It seems like it would be safe to reorder a load to be anywhere between the last place that the local was written to (which this pass enforces) and the first place that the load's lvalue is read from -- which I would expect is enforced by the existing dependency analysis. I'm sure there's something I'm missing here though!

Also -- if the thing we want to validate is just that the local is used exactly once, maybe rather than building an accessed range for the local, we could just count the number of reads?

Reply:
Yeah i was on the fence about checking the accessed range or just counting reads. Without some form of this check i saw some cases where a LoadLocal was used twice, but once was at the wrong block scope level, so it would be invalid to reorder to the first usage. I'll update this to just count reads and only reorder for LoadLocals whose temp is used exactly once.